### PR TITLE
[Merged by Bors] - feat: `Small.{u}` sets are closed under various operations

### PR DESCRIPTION
**Pull request by timotree3**
This PR adds lemmas showing that sets constructed by applying basic set operations to small sets are themselves small. We now have corresponding `Small ↥s` versions of almost all of the `Finite ↥s` instances in the `Finite.Set` namespace.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by timotree3** [comment]
!bench

File: Mathlib/Logic/Small/Basic.lean [hunk 1]
```diff
@@ -23,7 +23,11 @@ section
 
 open Classical
 
-instance small_subtype (α : Type v) [Small.{w} α] (P : α → Prop) : Small.{w} { x // P x } :=
+-- TODO(timotree3): lower the priority on this instance?
+-- This instance applies to every synthesis problem of the form `Small ↥s` for some set `s`,
+-- but we have lots of instances of `Small` for specific set constructions.
+instance small_subtype (α : Type v) [Small.{w} α] (P : α → Prop) :
```
Comment (urkud) [inline Mathlib/Logic/Small/Basic.lean:29]: I agree that this should have lower priority, but still higher than "generic" instances. Try `900`?

Comment (timotree3) [reply Mathlib/Logic/Small/Basic.lean]: Lowering the priority is causing compilation errors in later files that I don't know how to debug. Reverting for now


<details><summary>Example error</summary>

```text
[3216/4357]error: > LEAN_PATH=./.lake/packages/std/.lake/build/lib:./.lake/packages/Qq/.lake/build/lib:./.lake/packages/aesop/.lake/build/lib:./.lake/packages/proofwidgets/.lake/build/lib:./.lake/packages/Cli/.lake/build/lib:./.lake/packages/importGraph/.lake/build/lib:./.lake/build/lib LD_LIBRARY_PATH=./.lake/build/lib /home/lean/.elan/toolchains/leanprover--lean4---v4.7.0-rc2/bin/lean -Dpp.unicode.fun=true -Dpp.proofs.withType=false -DautoImplicit=false -DrelaxedAutoImplicit=false ./././Mathlib/Algebra/Module/Injective.lean -R ././. -o ./.lake/build/lib/Mathlib/Algebra/Module/Injective.olean -i ./.lake/build/lib/Mathlib/Algebra/Module/Injective.ilean -c ./.lake/build/ir/Mathlib/Algebra/Module/Injective.c
error: stdout:
Error: ./././Mathlib/Algebra/Module/Injective.lean:468:9: error: stuck at solving universe constraint
  v =?= max u ?u.297867
while trying to unify
  Shrink.{max u ?u.297867, u} ↥I →ₗ[R] ↥I : Type (max v u)
with
  LinearMap.{u, max u ?u.297867, u, u} (RingHom.id R) (Shrink.{max u ?u.297867, u} ↥I)
    ↥I : Type (max (max u ?u.297867) u)
Error: ./././Mathlib/Algebra/Module/Injective.lean:464:0: error: stuck at solving universe constraint
  v =?= max u ?u.297867
while trying to unify
  Shrink.{max u ?u.297867, u} ↥I →ₗ[R] ↥I : Type (max v u)
with
  LinearMap.{u, max u ?u.297867, u, u} (RingHom.id R) (Shrink.{max u ?u.297867, u} ↥I)
    ↥I : Type (max (max u ?u.297867) u)
error: external command `/home/lean/.elan/toolchains/leanprover--lean4---v4.7.0-rc2/bin/lean` exited with code 1
```

</details>

File: Mathlib/Logic/Small/Set.lean [hunk 2]
```diff
@@ -10,32 +10,108 @@ import Mathlib.Logic.Small.Basic
...
 
-instance small_range {α : Type v} {β : Type w} (f : α → β) [Small.{u} α] :
+instance small_powerset (s : Set α) [Small.{u} s] : Small.{u} (𝒫 s) :=
+  small_map (Equiv.Set.powerset s)
+
+instance small_sprod (s : Set α) (t : Set β) [Small.{u} s] [Small.{u} t] :
+    Small.{u} (s ×ˢ t : Set (α × β)) :=
+  small_of_injective (Equiv.Set.prod s t).injective
+
+instance small_set_pi {β : α → Type u2} (s : (a : α) → Set (β a))
```
Comment (YaelDillies) [inline Mathlib/Logic/Small/Set.lean:28]:
```suggestion
instance small_setPi {β : α → Type u2} (s : (a : α) → Set (β a))
```

File: Mathlib/Logic/Small/Set.lean [hunk 3]
```diff
@@ -10,32 +10,108 @@ import Mathlib.Logic.Small.Basic
...
+
+theorem small_subset {s t : Set α} (hts : t ⊆ s) [Small.{u} s] : Small.{u} t :=
+  small_of_injective (Set.inclusion_injective hts)
 #align small_subset small_subset
 
-instance small_range {α : Type v} {β : Type w} (f : α → β) [Small.{u} α] :
+instance small_powerset (s : Set α) [Small.{u} s] : Small.{u} (𝒫 s) :=
+  small_map (Equiv.Set.powerset s)
+
+instance small_sprod (s : Set α) (t : Set β) [Small.{u} s] [Small.{u} t] :
```
Comment (YaelDillies) [inline Mathlib/Logic/Small/Set.lean:24]:
```suggestion
instance small_setProd (s : Set α) (t : Set β) [Small.{u} s] [Small.{u} t] :
```

**Review by YaelDillies — APPROVED** [review]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada: 

If CI passes, please remove the label `awaiting-CI` and merge this yourself, by adding a comment `bors r+`.

bors d+

**Comment by robertylewis** [comment]
bors merge

File: Mathlib/Logic/Small/Set.lean [hunk 4]
```diff
@@ -10,32 +10,108 @@ import Mathlib.Logic.Small.Basic
...
+instance small_diff (s t : Set α) [Small.{u} s] : Small.{u} (s \ t : Set α) :=
+  small_subset (Set.diff_subset s t)
+
+instance small_sep (s : Set α) (P : α → Prop) [Small.{u} s] :
+    Small.{u} { x | x ∈ s ∧ P x} :=
+  small_subset (Set.sep_subset s P)
+
+instance small_inter_of_left (s t : Set α) [Small.{u} s] :
+    Small.{u} (s ∩ t : Set α) :=
+  small_subset (Set.inter_subset_left s t)
```
Comment (YaelDillies) [inline Mathlib/Logic/Small/Set.lean:77]:
```suggestion
  small_subset Set.inter_subset_left
```
and below

**Comment by jcommelin** [comment]
Please fix the mistakes and ping a reviewer/maintainer when you are done

bors d-

**Comment by vihdzp** [comment]
Are you still interested in this PR? If not, I can take it over.

**Comment by timotree3** [comment]
I believe all that was missing with this PR was fixing some easy merge conflicts. This should be ready for review if CI passes. Sorry for the long delay on my end!

**Comment by YaelDillies** [comment]
Okay great!

maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**